File: src/main/java/org/swordess/common/vitool/cmd/customize/Quit.java
package org.swordess.common.vitool.cmd.customize;

import org.springframework.shell.ExitRequest;
import org.springframework.shell.context.InteractionMode;
import org.springframework.shell.standard.ShellComponent;
import org.springframework.shell.standard.ShellMethod;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

@ShellComponent
public class Quit implements org.springframework.shell.standard.commands.Quit.Command {

    private static final List<Consumer<Void>> HANDLERS = new ArrayList<>();

    @ShellMethod(value = "Exit the shell.", key = {"quit", "exit"}, interactionMode = InteractionMode.INTERACTIVE,
            // place into the built-in group
            group = "Built-In Commands")
    public void quit() {
        HANDLERS.forEach(it -> it.accept(null));
        throw new ExitRequest();
    }

    public static void onExit(Consumer<Void> handler) {
        HANDLERS.add(handler);
    }

}

File: src/main/kotlin/org/swordess/common/vitool/cmd/customize/Quit.kt
/*
 * Copyright (c) 2019-2022 Swordess
 *
 * Distributed under MIT license.
 * See file LICENSE for detail or copy at https://opensource.org/licenses/MIT
 */

package org.swordess.common.vitool.cmd.customize

import org.springframework.shell.ExitRequest
import org.springframework.shell.context.InteractionMode
import org.springframework.shell.standard.ShellComponent
import org.springframework.shell.standard.ShellMethod
import org.springframework.shell.standard.commands.Quit


@ShellComponent
class Quit : Quit.Command {

    @ShellMethod(
        value = "Exit the shell.",
        key = ["quit", "exit"],
        interactionMode = InteractionMode.INTERACTIVE,
        group = "Built-In Commands"
    )
    fun quit() {
        fireExit()
        throw ExitRequest()
    }

    companion object {

        private var handlers: List<() -> Unit> = listOf()
        private var exited = false

        fun onExit(handler: () -> Unit) {
            handlers += handler
        }

        fun fireExit(): Boolean {
            if (exited) {
                return false
            }

            synchronized(Quit::class.java) {
                if (!exited) {
                    handlers.forEach { it.invoke() }
                    exited = true
                    return true
                }
            }

            return false
        }

    }

}